### DISCUSSION TITLE
v1.9 - Snapshot Data model refactor

### DISCUSSION THREAD

**Discussion by jwildfire** [Gilead-Public/gsm#1407] [General]
# Objective

Refactor the GSM snapshot data model to have a robust framework for longitudinal data storage and reporting. 

# Proposed Changes

- `make_snapshot()` gets a refector that incorporates `augment_snapshot` functionality. 
   - takes the current assessment (`lAssessment`) and previous snapshot (`lSnapshot_prev`) as inputs. 
       - `lSnapshot_prev` should be optional and return reasonable NAs for the first snapshot of the study
   - save all figures in the snapshot
   - incorporate `augment_snapshot` functionality in to `make_snapshot` then delete that function. 
- add new cumulative "result logs" to data model 
    - `results_summary_log` stacks all `result_summary` files from all snapshots. 
        - basically just `row_bind` `lSnapshot_prev$results_summary_log` and new `results_summary`
    - probably also need a `results_analysis_log` 
    - honestly, I guess we could create `{domain}_log` for all of the files in the snapshot model. Seems a little bit like overkill, but wouldn't be hard and would make us quite future proof ... 
- add `n_enrolled` to `snapshot$results_summary`
    - think through whether other columns are needed ... 
- add [`flag_history`](https://github.com/Gilead-BioStats/gsm/pull/1307) either as new domain (`results_flag_history`) or merged on to `results_summary`
- refactor all reports to use `lSnapshot` as input. 
- `StudyAssess` still makes `lAssessment`. It remains a list that has only cross-sectional data
   - Probably only needs minor changes

**Comment by mattroumaya** [comment]
- `lStackedSnapshots` will be created inside of `Make_Snapshot()` whenever a previous snapshot is provided.
- `lPrevSnapshot` should be a list object, imported from an .rda file, that contains historical/longitudinal data.
- Timeseries widgets need to be created inside of `Make_Snapshot()`, with logic for when longitudinal data is present.
  - May want to refactor the plots in general.
